### PR TITLE
test(kad): bootstrap and refactor find tests

### PR DESCRIPTION
**Pull request by rlve**
Changes:
- add bootstrap tests:
  - `"bootstrap with mixed fresh and stale buckets refreshes only stale"`
  - `"KadDHT Bootstrap Component - bootstrap discovers new peers through network"`
- refactor find tests
  - do not rely on bootstrap since we have dedicated bootstrap tests, connect nodes directly instead
  - use common helpers everywhere

### CODE REVIEW

File: tests/libp2p/kademlia/test_bootstrap.nim [hunk 1]
```diff
@@ -51,7 +42,7 @@ suite "KadDHT Bootstrap":
 
     # Add peers - they will be fresh (just added)
     kad.populateRoutingTable(5)
-    check kad.countNonEmptyBuckets() >= 1
+    check kad.getNonEmptyBuckets().len >= 1
```
Comment (gmelodie) [inline tests/libp2p/kademlia/test_bootstrap.nim]: nit
```suggestion
    check kad.nonEmptyBuckets().len >= 1
```

File: tests/libp2p/kademlia/test_bootstrap.nim [hunk 2]
```diff
@@ -65,35 +56,85 @@ suite "KadDHT Bootstrap":
...
+    check bucketIndices.len >= 2
+
+    for index in bucketIndices:
+      makeBucketStale(kad.rtable.buckets[index])
+
+    kad.findNodeCalls = @[]
+    await kad.bootstrap()
+
+    # Self lookup + one lookup per stale bucket
+    check kad.findNodeCalls.len == 1 + bucketIndices.len
```
Comment (gmelodie) [inline tests/libp2p/kademlia/test_bootstrap.nim]:
```suggestion
    check kad.findNodeCalls.len == bucketIndices.len + 1
```

File: tests/libp2p/kademlia/test_bootstrap.nim [hunk 3]
```diff
@@ -65,35 +56,85 @@ suite "KadDHT Bootstrap":
...
-      bucketIndex(kad.rtable.selfId, peerId.toKey(), kad.rtable.config.hasher)
-    kad.rtable.buckets[bucketIdx].peers[0].lastSeen = Moment.now() - 40.minutes
+    # Make only the first bucket stale
+    let staleBucketIndex = bucketIndices[0]
+    makeBucketStale(kad.rtable.buckets[staleBucketIndex])
+    check kad.rtable.buckets[staleBucketIndex].isStale()
 
-    check kad.rtable.buckets[bucketIdx].isStale()
+    # Verify that the rest of non-empty buckets is fresh
+    for index in bucketIndices[1 ..^ 1]:
```
Comment (gmelodie) [inline tests/libp2p/kademlia/test_bootstrap.nim]: does this `[1 ..^ 1]` notation not mean the same as [1]? If not, can you make it a bit clearer?

Comment (rlve) [reply tests/libp2p/kademlia/test_bootstrap.nim]: > Backwards index operator can be used in a slice to address the array from the high bound. ^1 is equivalent to arr1.len-1

So it means from 2nd element till the last. If you read the whole scenario it's pretty clear what it does.

Comment (gmelodie) [reply tests/libp2p/kademlia/test_bootstrap.nim]:
```nim
for i in 1 ..< bucketIndices.len:
  let index = bucketIndices[i]
```

Comment (rlve) [reply tests/libp2p/kademlia/test_bootstrap.nim]: You did not convince me 😄  This notation is used across the codebase, so I'll keep it

Comment (gmelodie) [reply tests/libp2p/kademlia/test_bootstrap.nim]: It's an ugly one, it shouldn't be used anywhere

Comment (vladopajic) [reply tests/libp2p/kademlia/test_bootstrap.nim]: sometimes we can fall into the trap of using the same method for everything. there are always edge cases when other methods should be used, or general principle shouldn't be followed.

since this is a test i wouldn't personally care about either, but if this is library code then variant with `[1 .. 1]` introduces unnecessary complexity (copying takes time and memory is trashed right after). 

personally `for i in 1 .. buckets.len` would be the preferred way  and if this could not do it, then `[ x .. y ]` as a last resort.

Comment (vladopajic) [reply tests/libp2p/kademlia/test_bootstrap.nim]: when writing code one should not use excuses not to do better, if legacy code was built with a suboptimal approach, one shouldn't really conform to it. a better code variant should always be preferred in new prs. this means that there will be different standards, but transitions are always painful.

File: tests/libp2p/kademlia/utils.nim [hunk 4]
```diff
@@ -145,6 +178,17 @@ proc getPeersfromRoutingTable*(kad: KadDHT): seq[PeerId] =
...
+proc nonEmptyBuckets*(kad: KadDHT): seq[int] =
+  var bucketIndices: seq[int]
+  for i, bucket in kad.rtable.buckets:
+    if bucket.peers.len > 0:
+      bucketIndices.add(i)
+  bucketIndices
+
+proc makeBucketStale*(bucket: var Bucket) =
+  for peer in bucket.peers.mitems:
+    peer.lastSeen = Moment.now() - 40.minutes
```
Comment (vladopajic) [inline tests/libp2p/kademlia/utils.nim]: 40.munutes must come from wherever this value is stored

File: tests/libp2p/kademlia/test_bootstrap.nim (hunk 3 above)
Comment (rlve) [reply tests/libp2p/kademlia/test_bootstrap.nim]: If notation `bucketIndices[1 ..^ 1]` is suboptimal due to performance reasons, then it's fine, I can change it. But this whole discussion started on it's being ugly, which is just an opinion

File: tests/libp2p/kademlia/utils.nim (hunk 4 above)
Comment (rlve) [reply tests/libp2p/kademlia/utils.nim]: https://github.com/vacp2p/nim-libp2p/blob/e468bf3bfb0d44afce3079f5191a9e19db93b7f4/libp2p/protocols/kademlia/routingtable.nim#L128-L134

Currently the threshold is not stored in any const.

Comment (vladopajic) [reply tests/libp2p/kademlia/utils.nim]: aha, but where does `30 minutes` come from? we should use that value here

Comment (vladopajic) [reply tests/libp2p/kademlia/utils.nim]: could we have constant for `30.minutes` and use it here?

Comment (rlve) [reply tests/libp2p/kademlia/utils.nim]: sure

File: tests/libp2p/kademlia/test_bootstrap.nim (hunk 3 above)
Comment (rlve) [reply tests/libp2p/kademlia/test_bootstrap.nim]: changed in https://github.com/vacp2p/nim-libp2p/pull/2023/changes/a65ae63a1bdc96251a53bf7a504d52003e339a11 🙄

File: tests/libp2p/kademlia/utils.nim (hunk 4 above)
Comment (rlve) [reply tests/libp2p/kademlia/utils.nim]: changed in https://github.com/vacp2p/nim-libp2p/pull/2023/changes/b23ecb88bbcffefc72399c62ff289989f0bb1209
